Fix signal handling with something that works

ggilmore opened the pull request:
The old signal handling code had a number of problems, mainly that it wasn't async-signal-safe (due to it calling a number nonreentrant) functions. See https://man7.org/linux/man-pages/man7/signal-safety.7.html for more information.

This PR solves the problem by simplifying our signal handling strategy:

 1. At the beginning of main (after initializing the P4API and de-registering its signal handlers), we block SIGINT, SIGHUP, SIGTERM, and SIGUSR1 in the main thread (and all the threads it spawns since signal masks are copied: see https://www.man7.org/linux/man-pages/man7/signal.7.html).

 2. We spawn a dedicated thread that uses sigwait (https://man7.org/linux/man-pages/man3/sigwait.3.html) to block until one of SIGINT, SIGHUP, SIGTERM, or SIGUSR1 is delivered. (We know that this is the only thread that receive these signals since we blocked it in every other thread in step 1). Once it receives on of these signals, it either:

	 - shuts down the thread pool cleanly and exits the process (if we receive a SIGINT, SIGTERM, or SIGHUP)
	
	 - just stops the signal handling thread (in the happy-path case where main tells the signal handler thread to shutdown (by sending it SIGUSR1) at the end of its execution

 The signal handling thread is allowed to use nonreentrant functions (like triggering condition variables for shutting down the threadpool) since it won't be interrupted by the arrival of another signal, etc.


## Test plan

- Here is a video of the the process cleanly shutting down if it doesn't receive any signals:

https://github.com/sourcegraph/p4-fusion/assets/9022011/b5cee8c8-110d-4963-a15f-448274d38574

- Here is a video of the process getting a SIGINT, and then the signal handling thread explicitly shutting down the threadpool

https://github.com/sourcegraph/p4-fusion/assets/9022011/a3808f7a-a085-4cca-872a-e8369257b8db

ggilmore committed:
Closes https://github.com/sourcegraph/sourcegraph/issues/57624

The old signal handling code had a number of problems, mainly that it wasn't async-signal-safe (due to it calling a number nonreentrant) functions. See https://man7.org/linux/man-pages/man7/signal-safety.7.html for more information.

This PR solves the problem by simplifying our signal handling strategy:

 1. At the beginning of main (after initializing the P4API and de-registering its signal handlers), we block SIGINT, SIGHUP, SIGTERM, and SIGUSR1 in the main thread (and all the threads it spawns since signal masks are shared: see https://www.man7.org/linux/man-pages/man7/signal.7.html).

 2. We spawn a dedicated thread that uses sigwait (https://man7.org/linux/man-pages/man3/sigwait.3.html) to block until one of SIGINT, SIGHUP, SIGTERM, or SIGUSR1 is delivered. (We know that this is the only thread that receive these signals since we blocked it in every other thread in step 1). Once it receives on of these signals, it either:

 - shuts down the thread pool cleanly and exits the process (if we receive a SIGINT, SIGTERM, or SIGHUP)

 - just stops the signal handling thread (in the happy-path case where main tells the signal handler thread to shutdown (by sending it SIGUSR1) at the end of its execution

 The signal handling thread is allowed to use nonreentrant functions (like triggering condition variables for shutting down the threadpool) since it won't be interrupted by the arrival of another signal, etc.

ggilmore commented on CMakeLists.txt:
```diff
@@ -12,12 +12,16 @@ option(THREADSAFE "Threadsafe" ON)
-    set(OPENSSL_ROOT_DIR "/usr/local/ssl/")
-endif()
+
+if (NOT DEFINED OPENSSL_ROOT_DIR)
```
Unrelated, but it was annoying we were essentially ignoring any custom value of OPENSSL_ROOT_DIR that the caller might have preferred to use.

eseliger commented on p4-fusion/main.cc:
```diff
@@ -63,9 +64,67 @@ int Main(int argc, char** argv)
+
+	// Create the thread pool
+	int networkThreads = arguments.GetNetworkThreads();
+	PRINT("Network Threads: " << networkThreads);
```
can we leave this where it was? It feels nice to get a summary of all the settings in a single place.

ggilmore replied:
Do you mean the log line itself, or the declaration of the thread pool? I wanted to start the signal handling thread as close to the invocation of pthread_sigmask as possible, which is why I needed the threadpool construction to move up too.

eseliger replied:
just this specific line 87 :)

eseliger commented on p4-fusion/main.cc:
```diff
@@ -63,9 +64,67 @@ int Main(int argc, char** argv)
+		    }
+
+		    ERR("(signal handler) received signal (" << sig << ") \"" << strsignal(sig) << "\", shutting down thread pool");
+		    pool.ShutDown();
```
The main thread does a few more things on shutdown, like making sure tracing data is written, and that the P4 API is properly shut down. Can we somehow do the same here?

eseliger replied:
From the main thread we're also hard-interrupting writing commits and trees now, since we call exit here.

I think in go I would finish writing the current changelist and the waitfordownload should be canceled with a context cancel error. Not sure how I would express that in C++.

Not a thing we have to solve in this PR, but feels a bit abrupt.

ggilmore replied:
> The main thread does a few more things on shutdown, like making sure tracing data is written, and that the P4 API is properly shut down. Can we somehow do the same here?

Done in [`43acb58` (#9)](https://github.com/sourcegraph/p4-fusion/pull/9/commits/43acb58f8c3beb1544d7d31621701478ba5c2de3)

ggilmore replied:
> From the main thread we're also hard-interrupting writing commits and trees now, since we call exit here.
> 
> I think in go I would finish writing the current changelist and the waitfordownload should be canceled with a context cancel error. Not sure how I would express that in C++.
> 
> Not a thing we have to solve in this PR, but feels a bit abrupt.

Maybe I'm missing something, but at a quick glance I would think this work would be encapsulated in a job (and will be cleanly shutdown when the thread pool is shutdown - but maybe I'm missing something)? We can chat on Monday about this though

eseliger commented:
cc @varungandhi-src

eseliger reviewed (approved):
great work digging into the weeds of signal handlers :)

ggilmore committed:
shutdown thread pool, P4API, and tracing in signal handler

varungandhi-src commented on p4-fusion/main.cc:
```diff
@@ -63,9 +64,77 @@ int Main(int argc, char** argv)
+	    {
+		    auto shutdown = [&pool]()
+		    {
+			    pool.ShutDown();
```
Not all the state inside the `ThreadPool` is protected by a mutex. For example, the `Shutdown` method itself starts mutating state immediately without taking a mutex.

So it's not obvious that calling this method is safe from a different thread. After all, later in this function, there is other code which may be concurrently calling other methods on the `ThreadPool`.

ggilmore replied:
You're correct. I saw that the shutdown code was already using atomic / condition variables, but I didn't think to check to see if the existing Shutdown logic was actually threadsafe / sound. After reviewing it, I found some issues and attempted to fix them in [`5ed3404` (#9)](https://github.com/sourcegraph/p4-fusion/pull/9/commits/5ed34043ae8c30c1fa3a00f3ad23ddc3d1cc0935) (see the commit message). PTAL?

varungandhi-src commented on p4-fusion/main.cc:
```diff
@@ -63,9 +64,77 @@ int Main(int argc, char** argv)
+
+		    // Wait for signals to arrive.
+		    int sig = 0;
+		    int rc = sigwait(&blockedSignals, &sig);
```
This is quite confusing to read because of the variable name `blockedSignals`, because in this sigwait call, it's not a "set by negation" but by inclusion. Moreover, it is being captured by (mutable) reference inside this thread, which creates a concern around race conditions/concurrent mutation.

I recommend instead creating a copy of that variable into a separate one while capturing it by value. Example code: https://godbolt.org/z/8sav5vM37

ggilmore replied:
Sure, happy to change the name and copy it. Done in [`9cf654e` (#9)](https://github.com/sourcegraph/p4-fusion/pull/9/commits/9cf654e4ea61b9973b0f90cd6986d2ae46c88e9a)

(Unfortunately, I had to use slightly uglier syntax since we can't initialize variables for closures in C++11.)

varungandhi-src commented:
Unrelated to the content of the PR but:

> (and all the threads it spawns since signal masks are shared: see [man7.org/linux/man-pages/man7/signal.7.html](https://www.man7.org/linux/man-pages/man7/signal.7.html)).

Nit: The use of "shared" here is confusing. AIUI, they're not actually shared, but inherited across thread spawns. "shared" would imply aliasing (and hence changing it in one thread retroactively would change it in other threads).

varungandhi-src commented on p4-fusion/main.cc:
```diff
@@ -63,9 +64,77 @@ int Main(int argc, char** argv)
+	auto signalHandlingThread = std::thread(
+	    [&blockedSignals, &pool]()
+	    {
+		    auto shutdown = [&pool]()
```
Suggestion: Factor out this `shutdown` function to be a top-level function which takes an exit code and a reference to the pool, have it call `std::exit()`. Ideally, all shutdown codepaths should be directed to the same shutdown function, but there is slightly different logic inside this `shutdown()` lambda compared to the end of `main` right now.

ggilmore replied:
[`d8159cb` (#9)](https://github.com/sourcegraph/p4-fusion/pull/9/commits/d8159cbfba53d64f535cc9fbc3215165c0e38821)

Do you mean this? I'm happy to make the change, but is there a stylistic / convention reason you're looking for this v. a lambda?

varungandhi-src replied:
> Do you mean this?

Yes, but also to replace this at the end of `main()`

```
	pool.ShutDown();
	// Wait for exception handler to finish.
	t.join();
	if (!P4API::ShutdownLibraries())
	{
		return 1;
	}
	// Finalize tracing.
	mtr_flush();
	mtr_shutdown();
```

With a call to the `shutdown()` function.

> but is there a stylistic / convention reason you're looking for this v. a lambda?

1. The problem with the lambda is that it's very tempting to use a capture `&pool` rather than taking it as an argument (which would introduce more syntactic noise at each call-site site). However, if it's captured by reference like that, then it's much less obvious if there's going to be concurrent mutation which may cause a data race.
2. The `main()` function is already long enough that it is a bit unwieldy to understand quickly, so separating out parts of the cleanup logic into a separate function helps with reasoning.

varungandhi-src commented:
> SIGINT, SIGHUP, SIGTERM, and SIGUSR1 [...]
>
> The signal handling thread is allowed to use nonreentrant functions (like triggering condition variables for shutting down the threadpool) since it won't be interrupted by the arrival of another signal, etc.

What is the expected behavior if a signal other than these signals is delivered to the process?

ggilmore commented:
> > SIGINT, SIGHUP, SIGTERM, and SIGUSR1 [...]
> > The signal handling thread is allowed to use nonreentrant functions (like triggering condition variables for shutting down the threadpool) since it won't be interrupted by the arrival of another signal, etc.
> 
> What is the expected behavior if a signal other than these signals is delivered to the process?

If another signal is delivered to the process, then the [default disposition](https://www.man7.org/linux/man-pages/man7/signal.7.html) for the signal will apply. (Put another way, if the default behavior is to ignore the signal, core dump, etc. - that action will still occur. The way we handle other signals is the same as it was before this change)

ggilmore committed:
thread-pool: make shutdown logic threadsafe

The old shutdown logic in threadpool wasn't threadsafe. Even though we use an atomic boolean to indicate a shutdown (m_HasShutDownBeenCalled), it was still possible for two threads both execute the Shutdown() routine at the same time (both threads could read m_HasShutDownBeenCalled == false at the top of the function). This was problematic since:

 1. The shutdown() routine join()'s the worker threads (it's UB if join() is called by multiple threads)
 2. The shutdown() routine clear()'s the job vector, but it was possible for other threads to be adding elements to the vector at the same time

 This PR attempts to improve things by:

 1. Introducing a new mutex for shutting down, m_ShutdownMutex, to prevent multiple threads from calling ShutDown() at the same time
 2. Enforcing that AddJob() holds the job mutex and checks to see if we have shutdown before adding a new element to the job queue to prevent simultaneous modification / memory leaks
 3. Introducing a new mutex for modifying the list of thread objects m_ThreadMutex for clarity (used for iterating over threads, populating threads, calculating the size of the number of threads in the threadpool, and clearing the list)
 4. Getting rid of m_ShouldStop since it was redundant with m_HasShutDownBeenCalled

ggilmore committed:
change variable name to signalsToWaitOn for signal handling thread

ggilmore commented:
> Unrelated to the content of the PR but:
> 
> > (and all the threads it spawns since signal masks are shared: see [man7.org/linux/man-pages/man7/signal.7.html](https://www.man7.org/linux/man-pages/man7/signal.7.html)).
> 
> Nit: The use of "shared" here is confusing. AIUI, they're not actually shared, but inherited across thread spawns. "shared" would imply aliasing (and hence changing it in one thread retroactively would change it in other threads).

You're correct - I used a poor word choice. The description now reads:

> ..., we block SIGINT, SIGHUP, SIGTERM, and SIGUSR1 in the main thread (and all the threads it spawns since signal masks are **copied** ...

ggilmore committed:
factor out shutdown in to top level function

ggilmore committed:
move "PRINT("Network Threads: " << networkThreads);" line to back where it was

varungandhi-src commented on p4-fusion/thread_pool.h:
```diff
@@ -39,16 +41,20 @@ class ThreadPool
 	ThreadPool(int size);
 	ThreadPool() = delete;
 	~ThreadPool();
 
-	void AddJob(Job function);
+	void AddJob(const Job& function);
```
The correct signature here should be `Job&&`, as the caller is expected to transfer the ownership of the Job to the threadpool, not to let the thread pool borrow the Job (which is what is implied by `const Job&`). Right now, the implementation of `AddJob` is performing a copy into the deque structure, which would be replaced by a `std::move(function)`.

varungandhi-src commented:
> Put another way, if the default behavior is to ignore the signal, core dump, etc. - that action will still occur. The way we handle other signals is the same as it was before this change)

The thing I'm wondering is, the OS will randomly chooses one of the application threads to handle the signal. Is it possible that if the thread that blocked on the `sigwait` is selected, then it might end up concurrently modifying some state along with some other thread? I believe that's not possible, because the default signal handling will apply, which shouldn't mutate any of the application state.

The other question is, if some other signal is delivered to terminate the application, do we make any kind of guarantees around the state on-disk (e.g. cloning should incrementally restart after termination at any point), or does one need to delete any incomplete state and start from scratch. I'm assuming that we're being careful about calling the cleanup routines properly for these specific signals to reduce/avoid the chances of leaving a partially cloned repository in a corrupt state.

ggilmore committed:
use shutdown helper routine in main() too

At the end of main(), we wait for the other threads to join() before we shut down. As such, we're not calling exit() in shutdown. We also can't include the thread join() routines in shutdown since those threads also call shutdown. As such, I thought the best solution was just to have each thread explicitly call exit() when they want to exit, and let main handle the thread killing, etc.

ggilmore committed:
use move semantics with AddJob

ggilmore commented:
> The thing I'm wondering is, the OS will randomly chooses one of the application threads to handle the signal. Is it possible that if the thread that blocked on the sigwait is selected, then it might end up concurrently modifying some state along with some other thread? I believe that's not possible, because the default signal handling will apply, which shouldn't mutate any of the application state.

You are correct. Amongst all the threads that haven't blocked the signal a thread will be picked at random by the OS. This includes our signal handling thread. However, we haven't registered a custom handler for any other signals, so the default disposition will apply (it won't run application code, the signal will be ignored, the process will be killed, etc. - whatever the default behavior is as specified in https://www.man7.org/linux/man-pages/man7/signal.7.html). There is no race condition since there won't be any application code running for these.

> The other question is, if some other signal is delivered to terminate the application, do we make any kind of guarantees around the state on-disk (e.g. cloning should incrementally restart after termination at any point), or does one need to delete any incomplete state and start from scratch. I'm assuming that we're being careful about calling the cleanup routines properly for these specific signals to reduce/avoid the chances of leaving a partially cloned repository in a corrupt state.

No, we make no guarantees about the on-disk state. Honestly, the signal that we're mostly likely to receive other than SIGINT is SIGKILL (if the process doesn't terminate quickly enough after a SIGINT for kubernetes' liking) - and it's not possible to catch SIGKILL anyway (your process will just get sniped).

With SIGINT (and similar) - we try to leave things in a somewhat consistent state (by shutting down the thread pool: which works by waiting for threads to complete their current work). However, we need to be robust against random process death anyway since that can happen for a variety of reasons (someone kicked the power plug, we get a SIGKILL, etc.). I don't think p4-fusion is this robust today (or at least isn't proven as such), but we can work to improve things there. (I also think gitserver's natural recloning behavior if it detects corruption can paper over a lot of things too)

ggilmore committed:
fix cmake
